File: annotation-processor-common/src/main/java/ru/tinkoff/kora/annotation/processor/common/FieldFactory.java
package ru.tinkoff.kora.annotation.processor.common;

import com.squareup.javapoet.*;

import javax.lang.model.element.Element;
import javax.lang.model.element.Modifier;
import javax.lang.model.type.TypeKind;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Types;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

public class FieldFactory {
    private final Types types;
    private final TypeSpec.Builder builder;
    private final MethodSpec.Builder constructor;
    private final Map<FieldKey, String> fields = new HashMap<>();
    private final String prefix;

    public String get(TypeName mapperType, Set<String> resultMapperTag) {
        return this.fields.get(new FieldKey(mapperType, resultMapperTag));
    }

    public String get(ClassName mapperType, TypeMirror mappedType, Element element) {
        var type = ParameterizedTypeName.get(mapperType, TypeName.get(mappedType).box());
        var tags = TagUtils.parseTagValue(element);
        var key = new FieldKey(type, tags);
        return this.fields.get(key);
    }

    public String get(TypeMirror mapperClass, Set<String> resultMapperTag) {
        var mapperType = TypeName.get(mapperClass);
        return this.fields.get(new FieldKey(mapperType, resultMapperTag));
    }

    record FieldKey(TypeName typeName, Set<String> tags) {}

    public FieldFactory(Types types, TypeSpec.Builder builder, MethodSpec.Builder constructor, String prefix) {
        this.types = types;
        this.builder = builder;
        this.constructor = constructor;
        this.prefix = prefix;
    }

    public String add(TypeName typeName, Set<String> tags) {
        var key = new FieldKey(typeName, tags);
        var existed = this.fields.get(key);
        if (existed != null) {
            return existed;
        }
        var name = this.prefix + (this.fields.size() + 1);
        this.fields.put(key, name);
        this.builder.addField(typeName, name, Modifier.PRIVATE, Modifier.FINAL);
        var parameter = ParameterSpec.builder(typeName, name);
        var tag = CommonUtils.toTagAnnotation(tags);
        if (tag != null) {
            parameter.addAnnotation(tag);
        }
        this.constructor.addParameter(parameter.build());
        this.constructor.addStatement("this.$N = $N", name, name);
        return name;
    }

    public String add(TypeName typeName, CodeBlock initializer) {
        var key = new FieldKey(typeName, Set.of());
        var existed = this.fields.get(key);
        if (existed != null) {
            return existed;
        }
        var name = this.prefix + (this.fields.size() + 1);
        this.fields.put(key, name);
        this.builder.addField(typeName, name, Modifier.PRIVATE, Modifier.FINAL);
        this.constructor.addStatement("this.$N = $L", name, initializer);
        return name;
    }

    public String add(TypeMirror typeMirror, Set<String> tags) {
        var typeName = TypeName.get(typeMirror);
        var key = new FieldKey(typeName, tags);
        var existed = this.fields.get(key);
        if (existed != null) {
            return existed;
        }
        var name = this.prefix + (this.fields.size() + 1);
        this.fields.put(key, name);
        this.builder.addField(typeName, name, Modifier.PRIVATE, Modifier.FINAL);
        if (tags.isEmpty() && typeMirror.getKind() == TypeKind.DECLARED && CommonUtils.hasDefaultConstructorAndFinal(this.types, typeMirror)) {
            this.constructor.addStatement("this.$N = new $T()", name, typeName);
        } else {
            this.constructor.addParameter(typeName, name);
            this.constructor.addStatement("this.$N = $N", name, name);
        }
        return name;
    }
}

File: symbol-processor-common/src/main/kotlin/ru/tinkoff/kora/ksp/common/FieldFactory.kt
package ru.tinkoff.kora.ksp.common

import com.google.devtools.ksp.getConstructors
import com.google.devtools.ksp.isOpen
import com.google.devtools.ksp.symbol.KSAnnotated
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSType
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterSpec.Companion.builder
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.ksp.toTypeName

class FieldFactory(builder: TypeSpec.Builder, constructor: FunSpec.Builder, prefix: String) {
    private val builder: TypeSpec.Builder
    private val constructor: FunSpec.Builder
    private val fields: MutableMap<FieldKey, String> = HashMap()
    private val prefix: String
    operator fun get(mapperType: TypeName, resultMapperTag: Set<String>): String? {
        return fields[FieldKey(mapperType, resultMapperTag)]
    }

    operator fun get(mapperType: ClassName, mappedType: KSType, element: KSAnnotated): String {
        val type = mapperType.parameterizedBy(mappedType.toTypeName().copy(false))
        val tags = TagUtils.parseTagValue(element)
        val key = FieldKey(type, tags)
        return fields[key]!!
    }

    operator fun get(mapperClass: KSType, resultMapperTag: Set<String>): String {
        val mapperType: TypeName = mapperClass.toTypeName()
        return fields[FieldKey(mapperType, resultMapperTag)]!!
    }

    internal data class FieldKey(val typeName: TypeName, val tags: Set<String>)

    init {
        this.builder = builder
        this.constructor = constructor
        this.prefix = prefix
    }

    fun add(typeName: TypeName, tags: Set<String>): String {
        val key = FieldKey(typeName, tags)
        val existed = fields[key]
        if (existed != null) {
            return existed
        }
        val name = prefix + (fields.size + 1)
        fields[key] = name
        builder.addProperty(name, typeName, KModifier.PRIVATE)
        val parameter = builder(name, typeName)

        if (tags.isNotEmpty()) {
            parameter.addAnnotation(AnnotationSpec.builder(CommonClassNames.tag).addMember("value = {%L}", tags.joinToString(", ") { "$it::class" }).build())
        }
        constructor.addParameter(parameter.build())
        constructor.addStatement("this.%N = %N", name, name)
        return name
    }

    fun add(typeName: TypeName, initializer: CodeBlock): String {
        val key = FieldKey(typeName, setOf())
        val existed = fields[key]
        if (existed != null) {
            return existed
        }
        val name = prefix + (fields.size + 1)
        fields[key] = name
        builder.addProperty(name, typeName, KModifier.PRIVATE)
        constructor.addStatement("this.%N = %L", name, initializer)
        return name
    }

    fun add(typeMirror: KSType, tags: Set<String>): String {
        val typeName = typeMirror.toTypeName()
        val key = FieldKey(typeName, tags)
        val existed = fields[key]
        if (existed != null) {
            return existed
        }
        val name = prefix + (fields.size + 1)
        fields[key] = name
        builder.addProperty(name, typeName, KModifier.PRIVATE)
        val decl = typeMirror.declaration
        if (tags.isEmpty() && decl is KSClassDeclaration && !decl.isOpen() && decl.getConstructors().count() == 1 && decl.getConstructors().first().parameters.isEmpty()) {
            constructor.addStatement("this.%N = %T()", name, typeName)
        } else {
            constructor.addParameter(name, typeName)
            constructor.addStatement("this.%N = %N", name, name)
        }
        return name
    }
}
